### DISCUSSION TITLE
Oracles

### DISCUSSION THREAD

**Discussion by marekkirejczyk** [noir-lang/noir#3767] [Ideas]
### Problem

We would like to create an Ethereum History API library, which would allow developers to access Ethereum data from within Noir circuit. The library would verify data with storage proofs. Data would be then submitted on-chain.

To do so, we would like to leverage unconstrained functions that would query an Ethereum node for data via Oracles. Oracles are mentioned in the Aztec documentation, but not yet implemented in pure Noir.

The goal is to reduce complexity for the developer, using the lib. With this approach, developers can manipulate data from Noir, without the need to write additional helper code in JS.

### Happy Case

Noir user pseudocode:
```rust
fn main(...) {
  …
  let storage = getStorage(address, blockNo)
  let result = … //some calculations on storage
  …
  result
}
```

User javascript pseudocode:
```js
const circuit = await getCircuit('main');
const zkProof = await noir.generateFinalProof({ params }, oracles);
```

Library javascript pseudocode:
```js
const oracles = {
    get_storage = () => {/* TODO */ return storage; },
    get_header = () => {/* TODO */ return header; },
    ....
}
```

Noir lib pseudocode:
```rust
#[oracle(get_storage)]
unconstrained fn get_storage(...) -> Storage {};
```

### Alternatives Considered

One common approach would be to fetch data from javascript and pass results to Noir code. 
This is a perfectly correct solution. However, it adds to complexity on the developer side. 
In this scenario, the developer needs to write code in 3 languages: js, Noir and Solidity.

### Would you like to submit a PR for this Issue?

No

**Comment by Savio-Sou** [comment]
Two possible approaches off the top of my head:

1. Create `get_storage_...` as a **special set of library functions** with enshrined tooling under the hood that fetches standardized information through standardized actions (e.g. a JSON-RPC call).

    Personally not a fan of enshrining special features.

2. Enable users to define **arbitrary functions paired with arbitrary embedded tooling**.

    Sounds like supporting users to run non-Noir code through Noir. Possibly achievable with additional macros or keywords.
